# Content-Based Filtering model  
Content-based filtering approaches leverage **description or attributes** from items the user has interacted to recommend similar items. It depends only on the user **previous choices**, making this method robust to **avoid the cold-start problem**. For textual items, like articles, news and books, it is simple to use the article **category** or **raw text** to build **item profiles** and **user profiles**.

Suppose I watch a particular genre movie I will be recommended movies w.r.t that specific genre. The Title, Year of Release, Director, Cast are also helpful in identifying similar movie content.

## Approach 2: Using Rated Content to Recommend
In this approach contents of the product are already **rated** and based on the **user’s preference**. An **item score** is predicted to products and recommendation can be made.

Usually `rating table` (user rating), `item profile` (book genres) are the only material we've got.

* `rating table`: user-to-book relationship
* `item profile`: attribute-to-book relationship  
![](https://www.analyticsvidhya.com/wp-content/uploads/2015/08/81.png)

Then we will create the `user profile` so that we can understand what attribute the users actually prefer.
* `user profile`: user-to-attribute relationship  
![](https://www.analyticsvidhya.com/wp-content/uploads/2015/08/91.png)

Thus, with the `user profile`, we can get all the item score which is the user preference from `user profile` and `item profile`.  
![](https://www.analyticsvidhya.com/wp-content/uploads/2015/08/121.png)

Let's go through the code!

In [1]:
import pandas as pd
import numpy as np
from tqdm import tqdm
from sklearn.preprocessing import MinMaxScaler
from merge_data import *

b'Skipping line 6452: expected 8 fields, saw 9\nSkipping line 43667: expected 8 fields, saw 10\nSkipping line 51751: expected 8 fields, saw 9\n'
b'Skipping line 92038: expected 8 fields, saw 9\nSkipping line 104319: expected 8 fields, saw 9\nSkipping line 121768: expected 8 fields, saw 9\n'
b'Skipping line 144058: expected 8 fields, saw 9\nSkipping line 150789: expected 8 fields, saw 9\nSkipping line 157128: expected 8 fields, saw 9\nSkipping line 180189: expected 8 fields, saw 9\nSkipping line 185738: expected 8 fields, saw 9\n'
b'Skipping line 209388: expected 8 fields, saw 9\nSkipping line 220626: expected 8 fields, saw 9\nSkipping line 227933: expected 8 fields, saw 11\nSkipping line 228957: expected 8 fields, saw 10\nSkipping line 245933: expected 8 fields, saw 9\nSkipping line 251296: expected 8 fields, saw 9\nSkipping line 259941: expected 8 fields, saw 9\nSkipping line 261529: expected 8 fields, saw 9\n'
/Users/Kevin/anaconda3/lib/python3.6/site-packages/IPython/core/interactiv

In [2]:
#import book csv with description, and more than 10 user ratings
df_n_des = pd.read_csv('books_n_description.csv', index_col=0)

In [3]:
df_n_des.head()

isbn                                              title  \
0  0002005018                                       Clara Callan   
1  0374157065  Flu: The Story of the Great Influenza Pandemic...   
2  0399135782                             The Kitchen God's Wife   
3  0440234743                                      The Testament   
4  0452264464               Beloved (Plume Contemporary Fiction)   

                 author  pub_year              publisher categories  \
0  Richard Bruce Wright    2001.0  HarperFlamingo Canada  Actresses   
1      Gina Bari Kolata    1999.0   Farrar Straus Giroux    Medical   
2               Amy Tan    1991.0       Putnam Pub Group    Fiction   
3          John Grisham    1999.0                   Dell    Fiction   
4         Toni Morrison    1994.0                  Plume    Fiction   

                                         description  
0  In a small town in Canada, Clara Callan reluct...  
1  Describes the great flu epidemic of 1918, an o...  
2  A Chinese immigrant who is convinced she is dy...  
3  A suicidal billionaire, a burnt-out Washington...  
4  Staring unflinchingly into the abyss of slaver...

In [4]:
df_n_des.shape

(15452, 7)

In [5]:
#make a copy to books_wd which categories cell is not null
books_wd = df_n_des[df_n_des['categories'].notnull()].copy()

# filter out books with less than 5 characters in categories
books_wd = books_wd[books_wd['categories'].map(len) >1]

In [6]:
books_wd.shape

(1466, 7)

We have **1,466** books in total

### Item profile

In [7]:
df_item = books_wd[['isbn','categories']]

In [8]:
df_item.head()

isbn categories
0  0002005018  Actresses
1  0374157065    Medical
2  0399135782    Fiction
3  0440234743    Fiction
4  0452264464    Fiction

In [9]:
# one-hot encoding for category
df_genre = pd.get_dummies(df_item['categories'])

let’s treat all articles as having unit weight.  
For binary representation, we can perform normalization by dividing the term occurrence by the sqrt of number of attributes in the article.  

In [10]:
#normalized
df_genre_normalized = df_genre.apply(lambda x: x/np.sqrt(df_genre.sum(axis=1)))

In [11]:
#create item profile
df_item = pd.concat([df_item, df_genre_normalized], axis=1)

df_item.drop(columns='categories', inplace=True)

In [12]:
df_item.sort_values('isbn', inplace=True)

In [13]:
df_item.set_index('isbn', inplace=True)

In [14]:
df_item.head()

Accidents  Action and adventure  Actors  Actresses  Adoptees  \
isbn                                                                       
0002005018        0.0                   0.0     0.0        1.0       0.0   
000648302X        0.0                   0.0     0.0        0.0       0.0   
000649840X        0.0                   0.0     0.0        0.0       0.0   
0020264763        0.0                   0.0     0.0        0.0       0.0   
0020264801        0.0                   0.0     0.0        0.0       0.0   

            Adventure stories  Affirmations  African American fiction  \
isbn                                                                    
0002005018                0.0           0.0                       0.0   
000648302X                0.0           0.0                       0.0   
000649840X                0.0           0.0                       0.0   
0020264763                0.0           0.0                       0.0   
0020264801                0.0           0.0                       0.0   

            African American men  African American psychologists  \
isbn                                                               
0002005018                   0.0                             0.0   
000648302X                   0.0                             0.0   
000649840X                   0.0                             0.0   
0020264763                   0.0                             0.0   
0020264801                   0.0                             0.0   

                   ...           Ryan, Jack (Fictitious character)  \
isbn               ...                                               
0002005018         ...                                         0.0   
000648302X         ...                                         0.0   
000649840X         ...                                         0.0   
0020264763         ...                                         0.0   
0020264801         ...                                         0.0   

            Savich, Dillon (Fictitious character)  Science fiction  Self-Help  \
isbn                                                                            
0002005018                                    0.0              0.0        0.0   
000648302X                                    0.0              0.0        0.0   
000649840X                                    0.0              0.0        0.0   
0020264763                                    0.0              0.0        0.0   
0020264801                                    0.0              0.0        0.0   

            Social Science  Star Trek fiction  Travel  Travelers' writings  \
isbn                                                                         
0002005018             0.0                0.0     0.0                  0.0   
000648302X             0.0                0.0     0.0                  0.0   
000649840X             0.0                0.0     0.0                  0.0   
0020264763             0.0                0.0     0.0                  0.0   
0020264801             0.0                0.0     0.0                  0.0   

            True Crime  Yorkshire (England)  
isbn                                         
0002005018         0.0                  0.0  
000648302X         0.0                  0.0  
000649840X         0.0                  0.0  
0020264763         0.0                  0.0  
0020264801         0.0                  0.0  

[5 rows x 158 columns]

In [15]:
df_item.shape

(1466, 158)

In `df_item`, there are **1,466** books and **158** genres.

Let's load the merged dataframe of `books`, `ratings`, `users` and get the record which matched the 1,466 books in `df_item`.

In [16]:
df_merged = merge_data_frame()

In [17]:
df_merged.head()

user        isbn  rating                           location   age  \
0  276725  034545104X       0                  tyler, texas, usa   NaN   
1    2313  034545104X       5              cincinnati, ohio, usa  23.0   
2    6543  034545104X       0           strafford, missouri, usa  34.0   
3    8680  034545104X       5  st. charles county, missouri, usa   2.0   
4   10314  034545104X       9             beaverton, oregon, usa   NaN   

  country   province                 title      author  pub_year  \
0     usa      texas  Flesh Tones: A Novel  M. J. Rose    2002.0   
1     usa       ohio  Flesh Tones: A Novel  M. J. Rose    2002.0   
2     usa   missouri  Flesh Tones: A Novel  M. J. Rose    2002.0   
3     usa   missouri  Flesh Tones: A Novel  M. J. Rose    2002.0   
4     usa     oregon  Flesh Tones: A Novel  M. J. Rose    2002.0   

          publisher                                              url_s  \
0  Ballantine Books  http://images.amazon.com/images/P/034545104X.0...   
1  Ballantine Books  http://images.amazon.com/images/P/034545104X.0...   
2  Ballantine Books  http://images.amazon.com/images/P/034545104X.0...   
3  Ballantine Books  http://images.amazon.com/images/P/034545104X.0...   
4  Ballantine Books  http://images.amazon.com/images/P/034545104X.0...   

                                               url_m  \
0  http://images.amazon.com/images/P/034545104X.0...   
1  http://images.amazon.com/images/P/034545104X.0...   
2  http://images.amazon.com/images/P/034545104X.0...   
3  http://images.amazon.com/images/P/034545104X.0...   
4  http://images.amazon.com/images/P/034545104X.0...   

                                               url_l  
0  http://images.amazon.com/images/P/034545104X.0...  
1  http://images.amazon.com/images/P/034545104X.0...  
2  http://images.amazon.com/images/P/034545104X.0...  
3  http://images.amazon.com/images/P/034545104X.0...  
4  http://images.amazon.com/images/P/034545104X.0...

In [18]:
#top users number and their number of ratings given
df_merged.user.value_counts().sort_values(ascending=False)[:10]

11676     11144
198711     6456
153662     5814
98391      5779
35859      5646
212898     4289
278418     3996
76352      3329
110973     2971
235105     2943
Name: user, dtype: int64

In [19]:
df_final = df_merged[df_merged['isbn'].isin(df_item.index)]

In [36]:
#create isbn and category dict
isbn_cat_dict = pd.Series(books_wd['categories'].values,index=books_wd['isbn']).to_dict()

#create isbn and description dict
isbn_des_dict = pd.Series(books_wd['description'].values,index=books_wd['isbn']).to_dict()

In [39]:
df_final['category'] = df_final['isbn'].map(isbn_cat_dict)
df_final['description'] = df_final['isbn'].map(isbn_des_dict)

/Users/Kevin/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.
/Users/Kevin/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


In [40]:
df_final.shape

(72821, 16)

In [42]:
df_final = df_final[['user','rating','isbn','title','author','publisher','pub_year','category','description']]

In [78]:
df_final.head()

user  rating        isbn                title     author  \
501  276746       0  0449006522  Manhattan Hunt Club  JOHN SAUL   
502  278026       8  0449006522  Manhattan Hunt Club  JOHN SAUL   
503     243       6  0449006522  Manhattan Hunt Club  JOHN SAUL   
504     645       0  0449006522  Manhattan Hunt Club  JOHN SAUL   
505    2010       0  0449006522  Manhattan Hunt Club  JOHN SAUL   

            publisher  pub_year category  \
501  Ballantine Books    2002.0  Fiction   
502  Ballantine Books    2002.0  Fiction   
503  Ballantine Books    2002.0  Fiction   
504  Ballantine Books    2002.0  Fiction   
505  Ballantine Books    2002.0  Fiction   

                                           description  
501  When college student Jeff Converse is wrongly ...  
502  When college student Jeff Converse is wrongly ...  
503  When college student Jeff Converse is wrongly ...  
504  When college student Jeff Converse is wrongly ...  
505  When college student Jeff Converse is wrongly ...

There are **72,821** ratings of the **1,466** books.

### Rating dataframe

In [44]:
rating = pd.pivot_table(df_final, values='rating', index=['isbn'], columns = ['user'])

In [45]:
rating.sort_index(axis=1, inplace=True)

In [46]:
rating.head()

user        10  100004  100009  10001  100029  100035  10005  100053  100066  \
isbn                                                                           
0002005018 NaN     NaN     NaN    NaN     NaN     NaN    NaN     NaN     NaN   
000648302X NaN     NaN     NaN    NaN     NaN     NaN    NaN     NaN     NaN   
000649840X NaN     NaN     NaN    NaN     NaN     NaN    NaN     NaN     NaN   
0020264763 NaN     NaN     NaN    NaN     NaN     NaN    NaN     NaN     NaN   
0020264801 NaN     NaN     NaN    NaN     NaN     NaN    NaN     NaN     NaN   

user        100067  ...    99885  99894  999  9991  99946  99955  99963  \
isbn                ...                                                   
0002005018     NaN  ...      NaN    NaN  NaN   NaN    NaN    NaN    NaN   
000648302X     NaN  ...      NaN    NaN  NaN   NaN    NaN    NaN    NaN   
000649840X     NaN  ...      NaN    NaN  NaN   NaN    NaN    NaN    NaN   
0020264763     NaN  ...      NaN    NaN  NaN   NaN    NaN    NaN    NaN   
0020264801     NaN  ...      NaN    NaN  NaN   NaN    NaN    NaN    NaN   

user        99973  99996  99997  
isbn                             
0002005018    NaN    NaN    NaN  
000648302X    NaN    NaN    NaN  
000649840X    NaN    NaN    NaN  
0020264763    NaN    NaN    NaN  
0020264801    NaN    NaN    NaN  

[5 rows x 22950 columns]

### User Profile

In [47]:
# get all the user number
users_no = rating.columns

In [48]:
#create an empty dataframe
df_users = pd.DataFrame(columns = df_item.columns)

In [49]:
df_users

Empty DataFrame
Columns: [Accidents, Action and adventure, Actors, Actresses, Adoptees, Adventure stories, Affirmations, African American fiction, African American men, African American psychologists, Age groups, Ahab, Captain (Fictitious character), Alchemists, Alcoholism, Altruism, Alzheimer's disease, American fiction, American literature, Americans, Andalusia (Spain), Animals, Anti-Catholicism, Antiques, Apocalyptic literature, Arabic fiction, Art, Artists, Astrology, Authors, American, Authors, Chilean, Baggins, Frodo (Fictitious character), Banks, Alan (Fictitious character), Bible, Biographical fiction, Biography &amp; Autobiography, Body, Mind &amp; Spirit, Book clubs (Discussion groups), Books and reading, Boston (Mass.), Boulder (Colo.), Boxcar children (Fictitious characters), Boys, Brain, Brooklyn (New York, N.Y.), Brunetti, Guido (Fictitious character), Business &amp; Economics, Business intelligence, Businessmen, Businesswomen, Calhoun, Mackenzie (Fictitious character), California, Canon (Literarature), Catskill Mountains Region (N.Y.), Childlessness, Children's stories, Comics &amp; Graphic Novels, Computers, Conduct of life, Confession, Continental European fiction (Fictional works by one author)., Cooking, Courtship, Covenant, Thomas (Fictitious character), Crafts &amp; Hobbies, Creative activities and seat work, Curiosities and wonders, Design, Detective and mystery stories, Diary fiction, Donation of organs, tissues, etc, Dracula, Count (Fictitious character), Dragons, Drama, Drug abuse, Drug traffic, Duitse fiksie, Education, Egypt, End of the world, England, English fiction, FICTION, Fairies, Fairy tales, Families, Family &amp; Relationships, Famines, Fantasy, Fantasy fiction, Female offenders, Feminists, Fiction, Fiction in English, Florida, Foreign Language Study, Fortune-telling by colors, France, Frankenstein (Fictitious character), Geishas, Generation X, ...]
Index: []

[0 rows x 158 columns]

In [50]:
for i in tqdm(range(len(users_no))):
    working_df = df_item.mul(rating.iloc[:,i], axis=0)
    working_df.replace(0, np.NaN, inplace=True)    
    df_users.loc[users_no[i]] = working_df.mean(axis=0)

100%|██████████| 22950/22950 [08:05<00:00, 47.30it/s]


In [54]:
#user profile
df_users.head()

Accidents  Action and adventure  Actors  Actresses  Adoptees  \
10            NaN                   NaN     NaN        NaN       NaN   
100004        NaN                   NaN     NaN        NaN       NaN   
100009        NaN                   NaN     NaN        NaN       NaN   
10001         NaN                   NaN     NaN        NaN       NaN   
100029        NaN                   NaN     NaN        NaN       NaN   

        Adventure stories  Affirmations  African American fiction  \
10                    NaN           NaN                       NaN   
100004                NaN           NaN                       NaN   
100009                NaN           NaN                       NaN   
10001                 NaN           NaN                       NaN   
100029                NaN           NaN                       NaN   

        African American men  African American psychologists  \
10                       NaN                             NaN   
100004                   NaN                             NaN   
100009                   NaN                             NaN   
10001                    NaN                             NaN   
100029                   NaN                             NaN   

               ...           Ryan, Jack (Fictitious character)  \
10             ...                                         NaN   
100004         ...                                         NaN   
100009         ...                                         NaN   
10001          ...                                         NaN   
100029         ...                                         NaN   

        Savich, Dillon (Fictitious character)  Science fiction  Self-Help  \
10                                        NaN              NaN        NaN   
100004                                    NaN              NaN        NaN   
100009                                    NaN              NaN        NaN   
10001                                     NaN              NaN        NaN   
100029                                    NaN              NaN        NaN   

        Social Science  Star Trek fiction  Travel  Travelers' writings  \
10                 NaN                NaN     NaN                  NaN   
100004             NaN                NaN     NaN                  NaN   
100009             NaN                NaN     NaN                  NaN   
10001              NaN                NaN     NaN                  NaN   
100029             NaN                NaN     NaN                  NaN   

        True Crime  Yorkshire (England)  
10             NaN                  NaN  
100004         NaN                  NaN  
100009         NaN                  NaN  
10001          NaN                  NaN  
100029         NaN                  NaN  

[5 rows x 158 columns]

### IDF
Let’s consider how common different terms are among our documents.  
The dot product of article vectors and IDF vectors gives us the **weighted scores** of each article.

In [103]:
document_frequency = df_item.sum()

In [104]:
idf = (len(books_wd)/document_frequency).apply(np.log) #log inverse of DF

In [108]:
#The dot product of article vectors and IDF vectors gives us the weighted scores of each article.
idf_df_item = df_item.mul(idf.values)

### Make Prediction

In [ ]:
#make an empty dataframe
df_predict = pd.DataFrame()

In [110]:
#user predict by tfidf
for i in tqdm(range(len(users_no))):
    working_df = idf_df_item.mul(df_users.iloc[i], axis=1)
    df_predict[users_no[i]] = working_df.sum(axis=1)

100%|██████████| 22950/22950 [04:18<00:00, 88.80it/s] 


In [111]:
#all user score predict of all books
df_predict.head()

10  100004    100009     10001  100029    100035  10005  \
isbn                                                                   
0002005018  0.0     0.0  0.000000  0.000000     0.0  0.000000    0.0   
000648302X  0.0     0.0  0.000000  0.000000     0.0  0.000000    0.0   
000649840X  0.0     0.0  0.000000  0.000000     0.0  0.000000    0.0   
0020264763  0.0     0.0  0.000000  0.000000     0.0  0.000000    0.0   
0020264801  0.0     0.0  2.777364  3.471705     0.0  2.777364    0.0   

              100053    100066    100067    ...     99885  99894       999  \
isbn                                        ...                              
0002005018  0.000000  0.000000  0.000000    ...       0.0    0.0  0.000000   
000648302X  0.000000  0.000000  0.000000    ...       0.0    0.0  0.000000   
000649840X  0.000000  0.000000  0.000000    ...       0.0    0.0  0.000000   
0020264763  0.000000  0.000000  0.000000    ...       0.0    0.0  0.000000   
0020264801  3.124534  2.198746  2.430193    ...       0.0    0.0  2.430193   

            9991     99946  99955  99963     99973     99996     99997  
isbn                                                                    
0002005018   0.0  0.000000    0.0    0.0  0.000000  0.000000  0.000000  
000648302X   0.0  0.000000    0.0    0.0  0.000000  0.000000  0.000000  
000649840X   0.0  0.000000    0.0    0.0  0.000000  0.000000  0.000000  
0020264763   0.0  0.000000    0.0    0.0  0.000000  0.000000  0.000000  
0020264801   0.0  1.909438    0.0    0.0  2.083023  3.008811  2.893087  

[5 rows x 22950 columns]

In [129]:
def recommender(user_no):
    
    #get all book isbn
    isbn_no = df_predict.index

    #user predicted rating to all books
    user_predicted_rating = df_predict['33570']

    #combine book rating and book detail
    user_rating_book = pd.concat([user_predicted_rating,books_wd.set_index('isbn')], axis=1)

    #books already read by user
    already_read = df_final[df_final['user'].isin(['33570'])]['isbn']

    #recommendation without books being read by user
    all_rec = user_rating_book[~user_rating_book.index.isin(already_read)]

    return all_rec.sort_values(by=[user_no], ascending=False).iloc[0:10]

In [130]:
recommender('33570')

/Users/Kevin/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:10: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  # Remove the CWD from sys.path while we load stuff.


33570                                              title  \
0449005569  2.430193                               Love: A User's Guide   
0449005410  2.430193          Horse Heaven (Ballantine Reader's Circle)   
0449130703  2.430193                            The Number of the Beast   
0449130509  2.430193                                       Winterbourne   
0449006689  2.430193  Murder in Havana (Truman, Margaret, Capital Cr...   
0449006344  2.430193                                        Angel Falls   
080411787X  2.430193                                       Acts of Love   
0804117934  2.430193   The Silent Cry (William Monk Novels (Paperback))   
0449004503  2.430193                                       Death Rounds   
0441003257  2.430193                                         Good Omens   

                     author  pub_year         publisher categories  \
0449005569     CLARE NAYLOR    1999.0  Ballantine Books    Fiction   
0449005410      Jane Smiley    2001.0  Ballantine Books    Fiction   
0449130703  Robert Heinlein    1989.0     Fawcett Books    Fiction   
0449130509    Susan Carroll    1987.0           Fawcett    Fiction   
0449006689  Margaret Truman    2002.0     Fawcett Books    Fiction   
0449006344   KRISTIN HANNAH    2001.0  Ballantine Books    Fiction   
080411787X   JUDITH MICHAEL    1997.0         Ivy Books    Fiction   
0804117934       Anne Perry    1998.0         Ivy Books    Fiction   
0449004503    PETER CLEMENT    1999.0           Fawcett    Fiction   
0441003257      Neil Gaiman    1996.0         Ace Books    Fiction   

                                                  description  
0449005569  When she catches the eye of London's handsomes...  
0449005410  A novel set in the world of thoroughbred racin...  
0449130703  The unusual adventures of four geniuses who ar...  
0449130509  Beloved author Susan Carroll took the romance ...  
0449006689  Asked to investigate an American pharmaceutica...  
0449006344  Liam will do anything to break his wife out of...  
080411787X  A collection of letters written by a young act...  
0804117934  Victorian-era criminal investigator William Mo...  
0449004503  The author, a former emergency room physician ...  
0441003257  When the armies of Heaven and Hell decide it's...

In [65]:
#books Read by user 33570
df_final[df_final['user'].isin(['33570'])].sort_values(by=['rating'], ascending=False)

user  rating        isbn  \
520     33570       8  0449006522   
188024  33570       6  0553583468   

                                               title      author  \
520                              Manhattan Hunt Club   JOHN SAUL   
188024  Whisper of Evil (Hooper, Kay. Evil Trilogy.)  Kay Hooper   

               publisher  pub_year category  \
520     Ballantine Books    2002.0  Fiction   
188024      Bantam Books    2002.0  Fiction   

                                              description  
520     When college student Jeff Converse is wrongly ...  
188024  As a series of grisly murders terrorizes the s...

### Pros
* **User independence**: collaborative filtering needs other users' rating to find the similarity between the users and then give the suggestion. Instead, content-based method only have to analyze the items and user profile for recommendation.
* **Transparency**: collaborative method gives you the recommendation because some unknown users have the same taste like you, but content-based method can tell you they recommend you the items based on what features. 
* **No cold start**: opposite to collaborative filtering, new items can be suggested before being rated by a substantial number of users.

### Cons
* **Limited content analysis**: if the content does not contain enough information to discriminate the items precisely, the recommendation will be not precisely at the end.
* **Over-specialization**: content-based method provides a limit degree of novelty, since it has to match up the features of profile and items. A totally perfect content-based filtering may suggest nothing "surprised." 
* **New user**: when there's not enough information to build a solid profile for a user, the recommendation could not be provided correctly. 


Reference  
https://www.analyticsvidhya.com/blog/2015/08/beginners-guide-learn-content-based-recommender-systems/  
https://towardsdatascience.com/learning-to-make-recommendations-745d13883951  
http://findoutyourfavorite.blogspot.com/2012/04/content-based-filtering.html